File: resources/ios/Sources/MobileBackgroundTasks/BackgroundTasksPlugin.swift
import Foundation
import BackgroundTasks

/**
 Background tasks bridge plugin.

 Uses iOS `BGTaskScheduler` to schedule periodic refresh and processing tasks.
 The actual Laravel command execution must be performed by your app's
 task handler — see the TODO blocks below.

 NOTE: iOS only allows task identifiers that are listed in
 `Info.plist > BGTaskSchedulerPermittedIdentifiers`. This plugin's
 `nativephp.json` declares `com.projectmata.task.*` as the allowed prefix.
 */
@objc(BackgroundTasksPlugin)
class BackgroundTasksPlugin: NSObject {

    fileprivate static var registered: [String: [String: Any]] = [:]

    @objc(BackgroundTasksPluginRegister)
    class Register: NSObject {
        @objc
        func execute(_ parameters: [String: Any]) -> [String: Any] {
            guard let tasks = parameters["tasks"] as? [[String: Any]] else {
                return [
                    "success": false,
                    "message": "tasks parameter is required."
                ]
            }

            for task in tasks {
                guard
                    let id = task["id"] as? String,
                    let intervalMinutes = task["intervalMinutes"] as? Int
                else { continue }

                let constraints = task["constraints"] as? [String: Any] ?? [:]
                BackgroundTasksPlugin.registered[id] = task

                let useProcessing = (constraints["longRunning"] as? Bool == true)
                    || (constraints["charging"] as? Bool == true)
                    || (constraints["idle"] as? Bool == true)

                let request: BGTaskRequest = useProcessing
                    ? BGProcessingTaskRequest(identifier: id)
                    : BGAppRefreshTaskRequest(identifier: id)

                if let processing = request as? BGProcessingTaskRequest {
                    processing.requiresExternalPower = (constraints["charging"] as? Bool) ?? false
                    processing.requiresNetworkConnectivity = (constraints["network"] as? String) != nil
                }

                if let refresh = request as? BGAppRefreshTaskRequest {
                    refresh.earliestBeginDate = Date(timeIntervalSinceNow: TimeInterval(intervalMinutes * 60))
                }

                do {
                    try BGTaskScheduler.shared.submit(request)
                } catch {
                    return [
                        "success": false,
                        "message": "Failed to submit task \(id): \(error.localizedDescription)"
                    ]
                }
            }

            return [
                "success": true,
                "registered": tasks.count
            ]
        }
    }

    @objc(BackgroundTasksPluginRunNow)
    class RunNow: NSObject {
        @objc
        func execute(_ parameters: [String: Any]) -> [String: Any] {
            // TODO: invoke each registered task's handler synchronously for testing.
            return [
                "success": true,
                "message": "RunNow dispatched (testing only)."
            ]
        }
    }

    @objc(BackgroundTasksPluginCancel)
    class Cancel: NSObject {
        @objc
        func execute(_ parameters: [String: Any]) -> [String: Any] {
            guard let taskId = parameters["taskId"] as? String else {
                return [
                    "success": false,
                    "message": "taskId is required."
                ]
            }

            BGTaskScheduler.shared.cancel(taskRequestWithIdentifier: taskId)
            BackgroundTasksPlugin.registered.removeValue(forKey: taskId)

            return [
                "success": true,
                "taskId": taskId
            ]
        }
    }

    @objc(BackgroundTasksPluginGetRegistered)
    class GetRegistered: NSObject {
        @objc
        func execute(_ parameters: [String: Any]) -> [String: Any] {
            let semaphore = DispatchSemaphore(value: 0)
            var pending: [[String: Any]] = []

            BGTaskScheduler.shared.getPendingTaskRequests { requests in
                pending = requests.map { request in
                    [
                        "id": request.identifier,
                        "earliestBeginDate": request.earliestBeginDate?.timeIntervalSince1970 ?? 0
                    ]
                }
                semaphore.signal()
            }

            semaphore.wait()

            return [
                "success": true,
                "tasks": pending
            ]
        }
    }
}

File: resources/android/src/main/java/com/projectmata/mobilebackgroundtasks/BackgroundTasksPlugin.kt
package com.projectmata.mobilebackgroundtasks

import androidx.fragment.app.FragmentActivity
import androidx.work.Constraints
import androidx.work.ExistingPeriodicWorkPolicy
import androidx.work.NetworkType
import androidx.work.PeriodicWorkRequestBuilder
import androidx.work.WorkManager
import com.nativephp.mobile.bridge.BridgeFunction
import com.nativephp.mobile.bridge.BridgeResponse
import com.nativephp.mobile.bridge.BridgeError
import java.util.concurrent.TimeUnit

/**
 * Background tasks bridge plugin.
 *
 * Uses WorkManager periodic work to run tasks at the requested cadence.
 * The actual execution of each `command` (e.g. `sync:data`) must be wired
 * up in your app's `BackgroundTasksWorker` — see the TODO blocks below.
 */
class BackgroundTasksPlugin {

    companion object {
        private fun makeError(code: String, message: String): BridgeError {
            val ctor = BridgeError::class.java.getDeclaredConstructor(
                String::class.java,
                String::class.java
            )
            ctor.isAccessible = true
            return ctor.newInstance(code, message)
        }

        private fun buildConstraints(map: Map<String, Any?>): Constraints {
            val builder = Constraints.Builder()

            when (map["network"]) {
                "any" -> builder.setRequiredNetworkType(NetworkType.CONNECTED)
                "wifi" -> builder.setRequiredNetworkType(NetworkType.UNMETERED)
                else -> builder.setRequiredNetworkType(NetworkType.NOT_REQUIRED)
            }

            if (map["charging"] == true) builder.setRequiresCharging(true)
            if (map["batteryNotLow"] == true) builder.setRequiresBatteryNotLow(true)
            if (map["storageNotLow"] == true) builder.setRequiresStorageNotLow(true)
            if (map["idle"] == true) builder.setRequiresDeviceIdle(true)

            return builder.build()
        }
    }

    class Register(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            return try {
                @Suppress("UNCHECKED_CAST")
                val tasks = parameters["tasks"] as? List<Map<String, Any>> ?: emptyList()

                val workManager = WorkManager.getInstance(activity)

                tasks.forEach { task ->
                    val taskId = task["id"]?.toString() ?: return@forEach
                    val command = task["command"]?.toString() ?: return@forEach
                    val intervalMinutes = (task["intervalMinutes"] as? Number)?.toLong() ?: 15L

                    @Suppress("UNCHECKED_CAST")
                    val constraintsMap = task["constraints"] as? Map<String, Any?> ?: emptyMap()

                    // TODO: replace BackgroundTasksWorker::class with your app's worker
                    // that knows how to invoke the Laravel `command` on the embedded runtime.
                    val request = PeriodicWorkRequestBuilder<BackgroundTasksWorker>(
                        intervalMinutes, TimeUnit.MINUTES
                    )
                        .setConstraints(buildConstraints(constraintsMap))
                        .addTag(taskId)
                        .setInputData(
                            androidx.work.workDataOf(
                                "command" to command,
                                "taskId" to taskId
                            )
                        )
                        .build()

                    workManager.enqueueUniquePeriodicWork(
                        taskId,
                        ExistingPeriodicWorkPolicy.UPDATE,
                        request
                    )
                }

                BridgeResponse.success(
                    mapOf<String, Any>(
                        "success" to true,
                        "registered" to tasks.size
                    )
                )
            } catch (e: Exception) {
                BridgeResponse.error(
                    makeError("BG_TASKS_REGISTER_ERROR", e.message ?: "Failed to register background tasks.")
                )
            }
        }
    }

    class RunNow(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            // TODO: dispatch each registered worker via OneTimeWorkRequest for testing.
            return BridgeResponse.success(
                mapOf<String, Any>(
                    "success" to true,
                    "message" to "RunNow dispatched (testing only)."
                )
            )
        }
    }

    class Cancel(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            return try {
                val taskId = parameters["taskId"]?.toString()
                    ?: return BridgeResponse.error(
                        makeError("BG_TASKS_BAD_ID", "taskId is required.")
                    )

                WorkManager.getInstance(activity).cancelUniqueWork(taskId)

                BridgeResponse.success(
                    mapOf<String, Any>(
                        "success" to true,
                        "taskId" to taskId
                    )
                )
            } catch (e: Exception) {
                BridgeResponse.error(
                    makeError("BG_TASKS_CANCEL_ERROR", e.message ?: "Failed to cancel task.")
                )
            }
        }
    }

    class GetRegistered(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            return try {
                val workManager = WorkManager.getInstance(activity)
                val infos = workManager.getWorkInfosByTag("com.projectmata.task").get()

                val tasks = infos.map { info ->
                    mapOf<String, Any>(
                        "id" to info.tags.firstOrNull { it.startsWith("com.projectmata.task") }.orEmpty(),
                        "state" to info.state.name
                    )
                }

                BridgeResponse.success(
                    mapOf<String, Any>(
                        "success" to true,
                        "tasks" to tasks
                    )
                )
            } catch (e: Exception) {
                BridgeResponse.error(
                    makeError("BG_TASKS_LIST_ERROR", e.message ?: "Failed to list registered tasks.")
                )
            }
        }
    }
}

/**
 * Stub worker — replace with the real implementation in your host app.
 * It must know how to invoke a Laravel artisan command on the embedded
 * NativePHP runtime and respect the `command` / `taskId` input data.
 */
class BackgroundTasksWorker(
    appContext: android.content.Context,
    workerParams: androidx.work.WorkerParameters
) : androidx.work.Worker(appContext, workerParams) {

    override fun doWork(): Result {
        val command = inputData.getString("command") ?: return Result.failure()
        // TODO: invoke `php artisan $command` against the bundled PHP runtime.
        return Result.success()
    }
}
